### PR TITLE
Add ability to customize CMake build directory

### LINKED ISSUE #19: Add an option to set relative build path
It should be configurable to set path for build folder. For example '..' (parent directory of the project root)

```
/Users/user/project/
├── Debug
    └── compile_commands.json
└── project
    ├── .git
    ├── CMakeLists.txt
    ├── compile_commands.json -> ../Debug/compile_commands.json
    └── src
        └── main.cpp
```

### PR DESCRIPTION
**Pull request by foonathan**
Fixes #19.

### CODE REVIEW

File: autoload/cmake/switch.vim
Comment (cdelledonne) [inline autoload/cmake/switch.vim:81]: This doesn't work in all cases unfortunately. When running CMake from e.g. a subdirectory of the project root (which Vim-CMake supports as of 743e3d316db3d3f08e348973c4e9c6807659c954), `build_dir` will be relative to the project root, and not to the CWD.

I would suggest to compute `l:build_dir` (which we should probably rename to `l:build_dir_parent`) the same as we compute `s:source_dir` in autoload/cmake.vim. When the build dir parent is not below the current dir, it will expand to a full path, not a relative path, but It doesn't matter in my opinion.

Let's also rename the function to `GetBuildDirParent()`, in light of the other comment I left in plugin/cmake.vim, and let's make this function "private", as we only have to access it from within this file.

File: autoload/cmake/switch.vim
Comment (cdelledonne) [inline autoload/cmake/switch.vim:17]: Perhaps `build_dir` -> `build_dir_parent`?`

**Review by cdelledonne — CHANGES_REQUESTED** [review]
I left a few comments, let me know what you think, thanks! :)

**Comment by cdelledonne** [comment]
By the way, no need to force-push your branch, we can always squash commits when merging. Actually separating commits might help pin-pointing the changes that were made after a review.

**Comment by karlosp** [comment]
@foonathan thanks for this patch. I hope it will be merged. For now, I am using your branch 👍 

I was surprised that by default it makes the build directory in-source because README says `The built files will end up in the binary directory (out-of-source build). `

Maybe one suggestion if it is possible to make interface something like this `:CMakeGenerate <config> <cmake_build_directory>`?

So one can specify <config> and overrides g:cmake_default_config and one can specify <cmake_build_directory> and overrides g:cmake_build_directory.

**Comment by cdelledonne** [comment]
@karlosp I think you might have misunderstood what "out-of-source build" means, the README has a link to a post that explains it, here it is: https://cprieto.com/posts/2016/10/cmake-out-of-source-build.html

Out-of-source means that the build artifacts, including for instance the `CMakeCache.txt` file, are in a separate directory, but the directory can be a child of the root directory. E.g., if the root directory is `project–foo/`, your build artifacts can end up in `project-foo/Debug/`, and the cache file would be `project-foo/Debug/CMakeCache.txt`. In-source means that your build artifacts end up in your project root, not in a separate directory, i.e., your cache file would be `project-foo/CMakeCache.txt` (which is messy, because you'd potentially have build artifacts and source files in the same directory).

The goal of #19 (and thus this PR) is to specify a path where the build directory must be placed, which by default if the project root (so a Debug configuration would build in `project-foo/Debug/`). For instance, one could specify to place the build directory inside of `builds/`, relative to the project root, to have the build directory as `project-foo/builds/Debug/`.

I don't think one would want to place the build directory **above** the project root, or at the same level, but you could do that. Definitely not a smart choice if you have multiple projects at the same level, since there would be one `Debug/` directory for both of them
```
.
├── project-foo/
├── project-bar/
└── Debug/  # artifacts of project-foo or project-bar?
```

Is this more clear?

Regarding this PR, I cannot merge it yet, as there is an issue, as I explained in this comment: https://github.com/cdelledonne/vim-cmake/pull/21#discussion_r528184283. @foonathan said he would address the issue to make this PR good to merge, even if he is not using this plugin any longer (thanks for that, and feel free to change your mind of course).

**Comment by karlosp** [comment]
@ cdelledonne thanks for your detailed explanation!

I have two reasons why would one want to put the build directory above the project root:

1. Real example: you export C++ code from MATLAB Simulink. The project root folder is generated with unknown folder names inside i.e one folder name is always user`s username:
```bash
.
$username
2018
```
You want to create generic CMakeLists.txt that you can copy to that project, so you use `file(GLOB` ... Now because build is in the root it will pick up some additional *.cpp files inside causing compilation error.

2. Sometimes I want to put it outside project root so that does not pollute my git status.

**Comment by cdelledonne** [comment]
@karlosp I see, thanks for sharing your use cases. Regarding Git, it's quite common to add build directories to the `.gitignore`, but of course that's not a rule.

**Comment by cdelledonne** [comment]
Handled in 6318c72321d9be3ac9c1fc1dac88531069472d78

**Pull request closed without merging**
